avr-hal-generic/src/spi.rs

```diff
@@ -0,0 +1,186 @@
+                    // enable SPI
+                    w.spe().set_bit();
+                    // Set to primary mode
+                    w.mstr().set_bit();
+                    // set up data order control bit
+                    match self.settings.data_order {
+                        DataOrder::MostSignificantFirst => w.dord().clear_bit(),
+                        DataOrder::LeastSignificantFirst => w.dord().set_bit(),
+                    };
+                    // set up polarity control bit
+                    match self.settings.mode.polarity {
+                        spi::Polarity::IdleHigh => w.cpol().set_bit(),
+                        spi::Polarity::IdleLow => w.cpol().clear_bit(),
+                    };
+                    // set up phase control bit
+                    match self.settings.mode.phase {
+                        spi::Phase::CaptureOnFirstTransition => w.cpha().clear_bit(),
+                        spi::Phase::CaptureOnSecondTransition => w.cpha().set_bit(),
+                    };
+                    // set up clock rate control bit
+                    match self.settings.clock {
+                        SerialClockRate::OscfOver2 => w.spr().val_0x00(),
+                        SerialClockRate::OscfOver4 => w.spr().val_0x00(),
+                        SerialClockRate::OscfOver8 => w.spr().val_0x01(),
+                        SerialClockRate::OscfOver16 => w.spr().val_0x01(),
+                        SerialClockRate::OscfOver32 => w.spr().val_0x02(),
+                        SerialClockRate::OscfOver64 => w.spr().val_0x02(),
+                        SerialClockRate::OscfOver128 => w.spr().val_0x03(),
+                    }
+                });
+                // set up 2x clock rate status bit
+                self.peripheral.spsr.write(|w| match self.settings.clock {
+                    SerialClockRate::OscfOver2 => w.spi2x().set_bit(),
+                    SerialClockRate::OscfOver4 => w.spi2x().clear_bit(),
+                    SerialClockRate::OscfOver8 => w.spi2x().set_bit(),
+                    SerialClockRate::OscfOver16 => w.spi2x().clear_bit(),
+                    SerialClockRate::OscfOver32 => w.spi2x().set_bit(),
+                    SerialClockRate::OscfOver64 => w.spi2x().clear_bit(),
+                    SerialClockRate::OscfOver128 => w.spi2x().set_bit(),
+                });
+            }
+        }
+
+        /// FullDuplex trait implementation, allowing this struct to be provided to
+        /// drivers that require it for operation.  Only 8-bit word size is supported
+        /// for now.
+        impl $crate::hal::spi::FullDuplex<u8> for $Spi {
+            type Error = $crate::void::Void;
+
+            /// Sets up the device for transmission and sends the data
+            fn send(&mut self, byte: u8) -> $crate::nb::Result<(), Self::Error> {
+                self.write(byte);
+                self.block_until_transfer_complete();
+                Ok(())
+            }
```

Comment:
This implementation does not quite behave like expected by _embedded-hal_.  It returns an `nb::Result` which is supposed to denote non-blocking behavior but your implementation is actually blocking.  You need to do this similar to what I have done in the serial module and return `nb::Error::WouldBlock` if the shift-register is not yet empty.  Then, in the calling code, wrap the call to `.send()` into an `nb::block!()`.

https://github.com/Rahix/avr-hal/blob/28a6837326005dfed2267c22b2e6dc5e0c69d958/avr-hal-generic/src/serial.rs#L101-L115

Reply:
Improved the nb implementation, see the latest.